flink-runtime/src/main/java/org/apache/flink/runtime/io/network/buffer/NetworkBufferPool.java

```diff
@@ -205,14 +236,26 @@ public void recycleMemorySegments(Collection<MemorySegment> segments) throws IOE
 	}
 
 	private void recycleMemorySegments(Collection<MemorySegment> segments, int size) throws IOException {
+		CompletableFuture<?> toNotify = null;
 		synchronized (factoryLock) {
 			numTotalRequiredBuffers -= size;
 
-			availableMemorySegments.addAll(segments);
+			synchronized (availabilityHelper) {
+				if (availableMemorySegments.isEmpty() && !segments.isEmpty()) {
+					toNotify = availabilityHelper.getUnavailableToResetAvailable();
+				}
+				availableMemorySegments.addAll(segments);
+				if (toNotify != null) {
```

Comment:
It can always call `availabilityHelper.notifyAll();`, because even though this condition is true it does not mean above  `availabilityHelper.wait(2000);` is happening.